datafusion/core/src/datasource/physical_plan/csv.rs

```diff
@@ -270,14 +297,223 @@ impl CsvOpener {
+                    return Ok(index);
+                }
+                index += 1;
+            }
+            Err(e) => {
+                return Err(DataFusionError::IoError(e));
+            }
+        }
+    }
+}
+
+/// Returns the offset of the first newline in the object store range [start, end), or the end offset if no newline is found.
+async fn find_first_newline(
+    object_store: &Arc<dyn ObjectStore>,
+    location: &object_store::path::Path,
+    start_byte: usize,
+    end_byte: usize,
+) -> Result<usize> {
+    let options = GetOptions {
+        range: Some(Range {
+            start: start_byte,
+            end: end_byte,
+        }),
+        ..Default::default()
+    };
+
+    let offset = match object_store.get_opts(location, options).await? {
+        GetResult::File(_, _) => {
+            // Range currently is ignored for GetResult::File(...)
+            // Alternative get_range() will copy the whole range into memory, thus set a limit of
+            // max bytes to read to find the first newline
+            let max_line_length = 4096; // in bytes
+            let get_range_end_result = object_store
+                .get_range(
+                    location,
+                    Range {
+                        start: start_byte,
+                        end: std::cmp::min(start_byte + max_line_length, end_byte),
+                    },
+                )
+                .await;
+            let mut decoder_tail = Cursor::new(get_range_end_result?);
+            find_first_newline_bytes(&mut decoder_tail)?
+        }
+        GetResult::Stream(s) => {
+            let mut input = s.map_err(DataFusionError::from);
+            let mut buffered = Bytes::new();
+
+            let future_index = async move {
+                let mut index = 0;
+
+                loop {
+                    if buffered.is_empty() {
+                        match input.next().await {
+                            Some(Ok(b)) => buffered = b,
+                            Some(Err(e)) => return Err(e),
+                            None => return Ok(index),
+                        };
+                    }
+
+                    for byte in &buffered {
+                        if *byte == b'\n' {
+                            return Ok(index);
+                        }
+                        index += 1;
+                    }
+
+                    buffered.advance(buffered.len());
+                }
+            };
+            future_index.await?
+        }
+    };
+    Ok(offset)
+}
+
 impl FileOpener for CsvOpener {
+    /// Open a partitioned CSV file.
+    ///
+    /// If `file_meta.range` is `None`, the entire file is opened.
+    /// If `file_meta.range` is `Some(FileRange {start, end})`, this signifies that the partition
+    /// corresponds to the byte range [start, end) within the file.
+    ///
+    /// Note: `start` or `end` might be in the middle of some lines. In such cases, the following rules
```

Comment:
As I understand it, this code does potentially several object store requests to adjust the initial ranges based on where the end of the CSV lines actually fall:

# Initial situation
```
CSV data with the next newlines (\n) after range marked                             
                                                                                    
 ┌─────────────┬──┬────────────────────┬──┬────────────────┬──┬────────┬──┬────────┐
 │     ...     │\n│        ...         │\n│      ...       │\n│  ...   │\n│  ...   │
 └─────────────┴──┴────────────────────┴──┴────────────────┴──┴────────┴──┴────────┘
        ▲                    ▲                     ▲               ▲                
        └─────────────┬──────┴─────────────────────┴───────────────┘                
                      │                                                             
                      │                                                             
             Initial file_meta.range es                                             
```

# This PR
```
    This PR: adjust the ranges prior to IO start, via                                
  object store operations                                                          
                                                                                   
┌ ─ ─ ─ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 
┌─────────────┬──┼────────────────────┬──┬────────────────┬──┬────────┬──┬────────┤
│     ...     │\n│        ...         │\n│      ...       │\n│  ...   │\n│  ...   │
└─────────────┴──┼────────────────────┴──┴────────────────┴──┴────────┴──┴────────┤
│                │                       │                   │                     
 ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘
                                                                                   
 Partition 0       Partition 1            Partition 2       Partition 3            
 Read              Read                   Read              Read                   
```

This design has the nice property that each partition reads exactly the bytes it needs

This design has the downside that it requires several object store reads to find the newlines, and the overhead of each object store operation often is much larger than the overhead of reading extra data. 

# Alternate idea

Another approach that would reduce the number of object store requests would be to read past the initial range and stop at the next newline `\n` like this:

```
  Each partition reads *more* than its assigned ranged to find                     
  the trailing new line, and ignores everything afterwards                         
                                                                                   
                                                                                   
┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─                                                          
┌─────────────┬──┬───────┼────────────┬──┬────────────────┬──┬────────┬──┬────────┐
│     ...     │\n│        ...         │\n│      ...       │\n│  ...   │\n│  ...   │
└─────────────┴──┴───────┼────────────┴──┴────────────────┴──┴────────┴──┴────────┘
│                                                                                  
 ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘                                                         
  Partition 0                                                                      
  Read                                                                             
```

I think the tricky bit of this design would be to ensure enough extra data was read. Initially, maybe we could just pick something sufficiently large for most files, like 1MB and error if the next newline can't be found. As a follow on we could add some fanciness like make another object store request if necessary.

Reply:
These text charts look awesome 😮 How did you draw that?

Reply:
I use https://monodraw.helftone.com/ to make the ASCII art -- I haven't found a free alternative yet

Reply:
> # Alternate idea
> Another approach that would reduce the number of object store requests would be to read past the initial range and stop at the next newline `\n` like this:
> 
> ```
>   Each partition reads *more* than its assigned ranged to find                     
>   the trailing new line, and ignores everything afterwards                         
>                                                                                    
>                                                                                    
> ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─                                                          
> ┌─────────────┬──┬───────┼────────────┬──┬────────────────┬──┬────────┬──┬────────┐
> │     ...     │\n│        ...         │\n│      ...       │\n│  ...   │\n│  ...   │
> └─────────────┴──┴───────┼────────────┴──┴────────────────┴──┴────────┴──┴────────┘
> │                                                                                  
>  ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘                                                         
>   Partition 0                                                                      
>   Read                                                                             
> ```
> 
> I think the tricky bit of this design would be to ensure enough extra data was read. Initially, maybe we could just pick something sufficiently large for most files, like 1MB and error if the next newline can't be found. As a follow on we could add some fanciness like make another object store request if necessary.

@alamb 
...working on #8502; I noticed the same thing - so am I right; that this "issue" is yet to be resolved? 

One could benchmark the different approaches (this PR vs. alternative Idea); I like the idea of picking a "sufficient large enough range at the beginning" and trying to recover with an additional object_store request.

Reply:
> I benchmarked it against 60mil rows of NDJSON - with no difference compared to the other approach. I could not gain any performance. Maybe my implementation is to naive? Or the overhead for the object store requests is to small when working with the local Filesystem?

Yes, I think this is the fundamental observation -- multiple requests to an actual remote object store is quite costly (like 10s of ms minimum) and fetching anything less than several MB in one request is likely less efficient than a single large request.

There is more about object storage in  [Exploiting Cloud Object Storage for High-Performance Analytics](https://www.vldb.org/pvldb/vol16/p2769-durner.pdf) : VLDB 2023 paper about running high performance analytics in object stores.

Reply:
@alamb
Thanks for linking the paper.

So I'll guess the alternate approach is still worth pursuing?

I would proceed by testing the POC against a remote store and if this looks promising - I'd create a separate issue to discuss and refine the approach further?

Reply:
Sounds like a good plan to me